src/BasicBehaveEngine/decorators/LoggingDecorator.ts

```diff
@@ -0,0 +1,142 @@
+import {ADecorator} from "./ADecorator";
+import {BehaveEngineNode} from "../BehaveEngineNode";
+import {IBehaveEngine, IEventQueueItem} from "../IBehaveEngine";
+import {IInteractivityFlow} from "../../types/InteractivityGraph";
+import { cubicBezier, linearFloat, slerpFloat4 } from "../easingUtils";
+export class LoggingDecorator extends ADecorator {
+    addToLog: (line: string) => void;
+    world: any;
+
+    constructor(behaveEngine: IBehaveEngine, addToLog: (line: string) => void, world: any) {
+        super(behaveEngine);
+        this.addToLog = addToLog;
+        this.world = world;
+        this.behaveEngine.processAddingNodeToQueue = this.processAddingNodeToQueue;
+        this.behaveEngine.processExecutingNextNode = this.processExecutingNextNode;
+        this.behaveEngine.processNodeStarted = this.processNodeStarted;
+        this.behaveEngine.animateProperty = this.animateProperty;
+        this.behaveEngine.animateCubicBezier = this.animateCubicBezier;
+        this.behaveEngine.getWorld = this.getWorld;
+
+        this.registerKnownPointers();
+    }
+
+    processAddingNodeToQueue = (flow: IInteractivityFlow) => {
+        this.addToLog(`Adding ${JSON.stringify(flow)} flow to queue`);
+    }
+
+    processExecutingNextNode = (flow: IInteractivityFlow) => {
+        this.addToLog(`Executing ${JSON.stringify(flow)} flow`);
+    }
+
+    processNodeStarted = (node: BehaveEngineNode) => {
+        this.addToLog(`Running ${node.name}: input values: ${JSON.stringify(node.values)}, output flows: ${JSON.stringify(node.flows)}`)
+    }
+
+    getWorld = (): any => {
+        console.log(this.world)
+        return this.world;
+    }
+
+    registerJsonPointer = (jsonPtr: string, getterCallback: (path: string) => any, setterCallback: (path: string, value: any) => void, typeName: string, readOnly: boolean) => {
+        this.behaveEngine.registerJsonPointer(jsonPtr, getterCallback, setterCallback, typeName, readOnly);
+    };
+
+    animateCubicBezier = (path: string, p1: number[], p2: number[], initialValue: any, targetValue: any, duration: number, valueType: string, callback: () => void): void => {
+        this.behaveEngine.getWorldAnimationPathCallback(path)?.cancel();
+
+        const startTime = Date.now();
+        const interpolationInterval = setInterval(() => {
+            const elapsedDuration = (Date.now() - startTime) / 1000;
+            const t = Math.min(elapsedDuration / duration, 1);
+            const p = cubicBezier(t, {x: 0, y:0}, {x: p1[0], y:p1[1]}, {x: p2[0], y:p2[1]}, {x: 1, y:1});
+            if (valueType === "float3") {
+                const value = [linearFloat(p.y, initialValue[0], targetValue[0]), linearFloat(p.y, initialValue[1], targetValue[1]), linearFloat(p.y, initialValue[2], targetValue[2])]
+                this.behaveEngine.setPathValue(path, value);
+            } else if (valueType === "float4") {
+                const value = slerpFloat4(t, initialValue, targetValue);
```

Comment:
Not all `float4` values are quaternions. For example, `pbrMetallicRoughness. baseColorFactor`.